neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+"""
+Base for "network" and "node" operations on neural masses.
+"""
+import logging
+
+import numpy as np
+import symengine as se
+from jitcdde import t as time_vector
+from jitcdde import y as state_vector
+
+from .backend import BackendIntegrator
+from .constants import (
+    EXC,
+    MASS_NAME_STR,
+    NETWORK_CONNECTIVITY,
+    NETWORK_DELAYS,
+    NETWORK_NAME_STR,
+    NODAL_CONNECTIVITY,
+    NODAL_DELAYS,
+    NODE_NAME_STR,
+)
+from .neural_mass import NeuralMass
+
+
+class Node(BackendIntegrator):
+    """
+    Base class for all nodes within the whole-brain network.
+    """
+
+    name = ""
+    label = ""
+
+    # index of this node with respect to the whole network
+    index = None
+
+    # list of synchronisation variables that are used to compute input/output
+    # from different masses in this node, implemented as `jitcdde` helpers
+    sync_variables = []
+
+    # default network coupling - will be used if network coupling is None,
+    # usually used for isolated node
+    default_network_coupling = {}
+
+    # default output of the model - all nodes need to have this variable defined
+    default_output = None
+
+    def __init__(self, neural_masses):
+        """
+        :param neural_masses: list of neural masses in this node
+        :type neural_masses: list[NeuralMass]
+        """
+        assert all(isinstance(mass, NeuralMass) for mass in neural_masses)
+        self.masses = neural_masses
+        # number of state variables for whole node, i.e. sum of state variables
+        # for all masses
+        self.num_state_variables = sum([mass.num_state_variables for mass in self])
+        self.num_noise_variables = sum([mass.num_noise_variables for mass in self])
```

Comment:
thanks to `__len__` and `__getitem__` functions, both `Node` and `Network` are actually iterators... so 
```python
# this iterates all masses in a node
for mass in self:
    print(mass)
```
or
```python
# this iterates nodes in a network
for node in self:
    # and this iterates all masses in this particular node
    for mass in node:
        print(mass)
```

you can appreciate how nice this is when you look at `describe()` function, or `get_nested_params()`.. since this is all hierarchy, network is just iterator of nodes, and node is just iterator of masses...
you can also do something like
```python
# first node in network
print(net[0])
# last node in network
print(net[-1])
# first mass in first node
print(net[0][0])
```
etc..

Reply:
DAMN! This is so good